dowhy/causal_model.py

```diff
@@ -459,4 +460,48 @@ def summary(self, print_to_stdout=False):
             print(summary_text)
         return summary_text
 
+    def refute_graph(self, k= 1, method_name =None, test_set = None ):
+        """
+        Check if the dependencies in input graph matches with the dataset
+        :param k: number of covariates in set Z while testing conditional independence ( X ⫫ Y ) | Z 
+        :param method_name: name of method to test conditional independece in data
+        :param test_set: list of implications to be test input by the user in the format 
+        [(x,y,(z1,z2)),
+        (x,y, (z3,))
+        ]
+        : returns: an instance of GraphRefutation class
+        """
+
+        refuter = GraphRefutation(data = self._data, method_name= method_name)
+
+        if test_set is None:
+            all_nodes=list(self._graph.get_all_nodes(include_unobserved=False))
+            all_possible_combinations = list(combinations(all_nodes, 2))  #Generating sets of size 2 for different x and y
+            conditional_independences=[]
+            self.logger.info("The followed conditional independences are true for the input graph")
+            for combination in all_possible_combinations:  #Iterate over the unique 2-sized sets [x,y]
+                a=combination[0]
+                b=combination[1]
+                all_nodes.remove(a) #Remove x from list of all nodes, in order to skip it from set Z
+                all_nodes.remove(b) #Remove y from list of all nodes, in order to skip it from set Z
+                k_sized_lists=list(combinations(all_nodes,k)) #Create k-sized unique Z set from remaining list of all nodes
+                for k_list in k_sized_lists:
+                    if self._graph.check_dseparation([str(a)], [str(b)], k_list) == True :
+                        self.logger.info(" %s and %s are CI given %s ", a, b, k_list)
+                        conditional_independences.append([a, b, k_list])
+                all_nodes.append(a) #Add x back to list of all nodes
```

Comment:
is there a better way to do this? currenlty append and remove are done in a loop, so they are repeated a lot of times. Also remove will be O(n) since it searches through all items in the list. 
Two suggestions: 
1. Rather than looping through all possible combinations, loop through the `all_nodes` list. You can start with `all_nodes[0]` and `all_nodes[1]` as `a` and `b`. then you basically need a nested loop (just like bubble sort) to list all combinations. 
2.   create a new temporary list variable that is obtained after removing a and b. That way you don't need to add it back.